DataFormats/MuonDetId/interface/GEMDetId.h

```diff
@@ -8,147 +8,293 @@
-  int roll() const{
-    return int((id_>>RollStartBit_) & RollMask_); // value 0 is used as wild card
+  /** Layer id: each station have two layers of chambers for GE1 and GE2: 
+      layer 1 is the inner chamber and layer 2 is the outer chamber 
+      For ME0 there are 6 layers of chambers */
+  constexpr int layer() const {
+    return (static_cast<int>((id_>>LayerStartBit) & LayerMask) + minLayerId);
   }
 
-  /// Return the corresponding ChamberId
-  GEMDetId chamberId() const {
-    return GEMDetId(id_ & chamberIdMask_);
+  /** Roll id  (also known as eta partition): each chamber is divided along 
+      the strip direction in  several parts  (rolls) GEM up to 12 */
+  constexpr int roll() const {
+    return (static_cast<int>((id_>>RollStartBit) & RollMask)); // value 0 is used as wild card
   }
 
-  /// Return the corresponding superChamberId
-  GEMDetId superChamberId() const {
-    return GEMDetId(id_ & superChamberIdMask_);
+  /** Return the corresponding ChamberId */
+  constexpr GEMDetId chamberId() const {
+    return GEMDetId(id_ & chamberIdMask);
   }
 
-  static const int minRegionId=     -1;
-  static const int maxRegionId=      1;
- 
-  static const int minRingId=   1;
-  static const int maxRingId=   3;
-
-  static const int minStationId=     1;
-  static const int maxStationId=     2;  // in the detId there is space to go up to 4 stations. Only 2 implemented now
-
-  static const int minChamberId=     0;
-  static const int maxChamberId=     36;
-
-  // LayerId = 0 is superChamber
-  static const int minLayerId=     0;
-  static const int maxLayerId=     2;
-
-  static const int minRollId=	  0;
-  static const int maxRollId=	 15;
-
- private:
-  static const int RegionNumBits_  =  2;
-  static const int RegionStartBit_ =  0;  
-  static const int RegionMask_     =  0X3;
+  /** Return the corresponding superChamberId */
+  constexpr GEMDetId superChamberId() const {
+    return GEMDetId(id_ & superChamberIdMask);
+  }
 
-  static const int RingNumBits_  =  3;
-  static const int RingStartBit_ =  RegionStartBit_+RegionNumBits_;  
-  static const unsigned int RingMask_     =  0X7;
+  /** Return the corresponding LayerId (mask eta partition) */
+  constexpr GEMDetId layerId() const {
+    return GEMDetId(id_ & chamberIdMask);
+  }
 
-  static const int StationNumBits_  =  3;
-  static const int StationStartBit_ =  RingStartBit_+RingNumBits_;  
-  static const unsigned int StationMask_     =  0X7;
+  /** Return total # of layers for this type of detector */
+  constexpr int nlayers() const {
+    return ((station() == 0) ? maxLayerId0 : ((station() > maxStationId) ?
+					     0 : maxLayerId));
+  }
 
+  constexpr uint32_t newForm() const {
+    return newForm(id_);
+  }
 
-  static const int ChamberNumBits_  =  6;
-  static const int ChamberStartBit_ =  StationStartBit_+StationNumBits_;  
-  static const unsigned int ChamberMask_     =  0X3F;
+  constexpr static uint32_t newForm(const uint32_t& inpid) {
```

Comment:
perhaps couple it to a more static reference of class version v11 as old?
So that "new" doesn't have to repeat again in the future.